ouroboros_macro/src/parse.rs

```diff
@@ -161,10 +160,16 @@ pub fn parse_struct(def: &ItemStruct) -> Result<StructInfo, Error> {
                         remove_attrs.push(index);
                     }
                     if path.segments.first().unwrap().ident == "covariant" {
+                        if covariant.is_some() {
+                            panic!("TODO: Nice error, covariance specified twice.");
+                        }
                         covariant = Some(true);
                         remove_attrs.push(index);
                     }
                     if path.segments.first().unwrap().ident == "not_covariant" {
+                        if covariant.is_some() {
+                            panic!("TODO: Nice error, covariance specified twice.");
+                        }
```

Comment:
Actually, covariance could also be automatically detected _and_ explicitly specified. Not sure if we should error in this case, explicit annotation should probably take precedent. Also, with these errors, making the covariance detection more precise would be a breaking change.